fiemap/src/lib.rs

```diff
@@ -21,21 +22,41 @@ pub struct Fiemap {
     ended: bool,
 }
 
-/// Get fiemap for the path and return an iterator of extents
+/// Get fiemap for the path and return an iterator of extents.
+///
+/// Same as [`Fiemap::new_from_path`].
 pub fn fiemap<P: AsRef<Path>>(filepath: P) -> Result<Fiemap> {
-    let file = File::open(filepath)?;
-    let fd = file.as_raw_fd();
-    Ok(Fiemap {
-        _file: file, // keep file alive
-        fd,
-        fiemap: C_fiemap::new(),
-        cur_idx: 0,
-        size: 0,
-        ended: false,
-    })
+    Fiemap::new_from_path(filepath)
 }
 
 impl Fiemap {
+    /// Creates a new [`Self`] from any type that implements [`AsFd`].
+    ///
+    /// The lifetime of the underlying file is tied to the lifetime of
+    /// the [`Self`] instance, as the file descriptor will be closed
+    /// after the instance of [`Self`] is dropped.
+    pub fn new(fd: impl AsRawFd) -> Self {
+        let raw_fd = fd.as_raw_fd();
+        let file = unsafe { File::from_raw_fd(raw_fd) };
```

Comment:
I don't quite like this idea. We can turn `_file` into an `Option` so that we can use the fd directly. `RawFd` is unclear about who closes the fd. Maybe we can use `BorrowedFd` or `OwnedFd` so that it's clear to the type checker who should close the fd?

Reply:
I just checked that a `File` struct is just a wrapper of an `OwnedFd`. We can accept `File` and let the caller convert if it's needed. The `FIEMAP` only makes sense for files, so that sounds reasonable too.

Reply:
Yes, this is what I meant initially as well and suggested this as well. There is no issue in changing the API, the crate is lower than 1.0 version, is dead-simple, and we can add a changelog to explain how to migrate, while providing the old API for some time for compatibility, while showing a deprecated message.